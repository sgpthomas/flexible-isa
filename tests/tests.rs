use std::{
    fmt::Display,
    path::{Path, PathBuf},
};

use flexible_isa::{cli, run, HalideExprOp, InstructionSelect};
use itertools::Itertools;

// #[test]
// fn parse_tests() {
//     todo!()
// }

const BENCHMARKS_PATH: &str = "cache";
#[allow(unused)]
const BENCHMAKRS: &[&str] = &[
    "add",
    "average_pool",
    "blur3x3",
    "camera_pipe",
    "conv3x3a16",
    "conv3x3a32",
    "conv_nn",
    "depthwise_conv",
    "dilate3x3",
    // "elementwise",
    // "fully_connected",
    "gaussian3x3",
    "gaussian5x5",
    "gaussian7x7",
    "l2norm",
    "matmul",
    "max_pool",
    "mean",
    "median3x3",
    "mul",
    "sobel3x3",
    "softmax",
];

fn test_single_isa<S: AsRef<Path> + Display>(name: S) -> anyhow::Result<()> {
    let mut path = PathBuf::from(BENCHMARKS_PATH).join(&name);
    path.set_extension("stmt");

    let args: cli::Args = cli::Args::new(&[path]).learn(true);
    let isa = run(args)?;

    let instr_hist = InstructionSelect::from_recexpr(&isa.expressions);
    let hist: Vec<_> = instr_hist
        .iter()
        .filter_map(|(op, count)| {
            if let HalideExprOp::Instruction(i) = op {
                Some((i, count))
            } else {
                None
            }
        })
        // sort first by count, then by index
        .sorted_by_key(|(i, count)| (*count, *i))
        .map(|(i, count)| (i, count, isa.instructions[&((*i) as usize)].to_string()))
        .collect();

    insta::with_settings!({
        description => format!("generate isa for {name}"),
        omit_expression => true,
        snapshot_suffix => format!("{name}")
    }, {
        insta::assert_debug_snapshot!(hist)
    });

    Ok(())
}

#[test]
fn add_isa() -> anyhow::Result<()> {
    test_single_isa("add")
}

#[test]
fn average_pool_isa() -> anyhow::Result<()> {
    test_single_isa("average_pool")
}

#[test]
fn blur3x3_isa() -> anyhow::Result<()> {
    test_single_isa("blur3x3")
}

#[test]
fn camera_pipe_isa() -> anyhow::Result<()> {
    test_single_isa("camera_pipe")
}

#[test]
fn conv3x3a16_isa() -> anyhow::Result<()> {
    test_single_isa("conv3x3a16")
}

#[test]
fn conv3x3a32_isa() -> anyhow::Result<()> {
    test_single_isa("conv3x3a32")
}

#[test]
fn conv_nn_isa() -> anyhow::Result<()> {
    test_single_isa("conv_nn")
}

#[test]
fn depthwise_conv_isa() -> anyhow::Result<()> {
    test_single_isa("depthwise_conv")
}

#[test]
fn dilate3x3_isa() -> anyhow::Result<()> {
    test_single_isa("dilate3x3")
}

#[test]
fn gaussian3x3_isa() -> anyhow::Result<()> {
    test_single_isa("gaussian3x3")
}

#[test]
fn gaussian5x5_isa() -> anyhow::Result<()> {
    test_single_isa("gaussian5x5")
}

#[test]
fn gaussian7x7_isa() -> anyhow::Result<()> {
    test_single_isa("gaussian7x7")
}

#[test]
fn l2norm_isa() -> anyhow::Result<()> {
    test_single_isa("l2norm")
}

#[test]
fn matmul_isa() -> anyhow::Result<()> {
    test_single_isa("matmul")
}

#[test]
fn max_pool_isa() -> anyhow::Result<()> {
    test_single_isa("max_pool")
}

#[test]
fn mean_isa() -> anyhow::Result<()> {
    test_single_isa("mean")
}

#[test]
fn median3x3_isa() -> anyhow::Result<()> {
    test_single_isa("median3x3")
}

#[test]
fn mul_isa() -> anyhow::Result<()> {
    test_single_isa("mul")
}

#[test]
fn sobel3x3_isa() -> anyhow::Result<()> {
    test_single_isa("sobel3x3")
}

#[test]
fn softmax_isa() -> anyhow::Result<()> {
    test_single_isa("softmax")
}
