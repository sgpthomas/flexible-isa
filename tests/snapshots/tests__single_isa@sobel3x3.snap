---
source: tests/tests.rs
description: generate isa for sobel3x3
---
[
    (
        4,
        1,
        "(+ (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> ?x1))) (cast<int32> 127))",
    ),
    (
        9,
        1,
        "(+ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> 127))",
    ),
    (
        10,
        1,
        "(+ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> (neg (cast<int32> 1))))",
    ),
    (
        12,
        1,
        "(+ (cast<int32> (- (cast<int32> (- (cast<int32> ?x0) (cast<int32> ?#2))) (cast<int32> ?x1))) (cast<int32> ?#3))",
    ),
    (
        14,
        1,
        "(+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> (neg (cast<int32> 1))))) (cast<int32> ?#2))) (cast<int32> ?x1))",
    ),
    (
        29,
        1,
        "(+ (cast<int32> (call[max] (cast<int32> (- (cast<int32> (call[min] (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?#4))) (cast<int32> ?x2))) (cast<int32> ?x3))) (cast<int32> 1))",
    ),
    (
        31,
        1,
        "(+ (cast<int32> (call[max] (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        33,
        1,
        "(+ (cast<int32> (call[min] (cast<int32> ?x0) (cast<int32> (+ (cast<int32> ?x1) (cast<int32> ?x2))))) (cast<int32> (neg (cast<int32> ?x3))))",
    ),
    (
        41,
        1,
        "(+ (cast<int32> ?x0) (cast<int32> 127))",
    ),
    (
        43,
        1,
        "(+ (cast<int32> ?x0) (cast<int32> (neg (cast<int32> 1))))",
    ),
    (
        55,
        1,
        "(- (cast<int32> (- (cast<int32> ?x0) (cast<int32> ?#2))) (cast<int32> ?x1))",
    ),
    (
        72,
        1,
        "(/ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        74,
        1,
        "(/ (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        80,
        1,
        "(get ?#1 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        81,
        1,
        "(get ?#1 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 130))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        88,
        1,
        "(get ?#1 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 512))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        89,
        1,
        "(get ?#1 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 514))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        91,
        1,
        "(get ?#3 (cast<int32x128> (+ (cast<int32x128> (call[max] (cast<int32x128> (call[min] (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> ?#4))) (cast<int32> ?x1))) (cast<int32> 1) (cast<int32> 128))) (cast<int32x128> (call[x128] (cast<int32> (+ (cast<int32> ?#5) (cast<int32> (neg (cast<int32> 1))))))))) (cast<int32x128> (call[x128] (cast<int32> 0))))) (cast<int32x128> (call[x128] (cast<int32> ?x2))))))",
    ),
    (
        0,
        2,
        "(neg (cast<int32> ?x0))",
    ),
    (
        3,
        2,
        "(+ (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 4))) (cast<int32> ?#2))) (cast<int32> ?x1))",
    ),
    (
        5,
        2,
        "(+ (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> ?x1))) (cast<int32> 128))",
    ),
    (
        35,
        2,
        "(+ (cast<int32> ?x0) (cast<int32> 2))",
    ),
    (
        36,
        2,
        "(+ (cast<int32> ?x0) (cast<int32> 3))",
    ),
    (
        40,
        2,
        "(+ (cast<int32> ?x0) (cast<int32> 13))",
    ),
    (
        42,
        2,
        "(+ (cast<int32> ?x0) (cast<int32> 128))",
    ),
    (
        44,
        2,
        "(+ (cast<int32> ?x0) (cast<int32> (neg (cast<int32> ?x1))))",
    ),
    (
        54,
        2,
        "(- (cast<int32> 0) (cast<int32> ?x0))",
    ),
    (
        59,
        2,
        "(- (cast<int32> ?x0) (cast<int32> (call[max] (cast<int32> (call[min] (cast<int32> (+ (cast<int32> ?x1) (cast<int32> ?x2))) (cast<int32> ?#4))) (cast<int32> ?x3))))",
    ),
    (
        62,
        2,
        "(* (cast<int32> (+ (cast<int32> (call[max] (cast<int32> (call[min] (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?#2))) (cast<int32> 1))) (cast<int32> (neg (cast<int32> 1))))) (cast<int32> ?#3))",
    ),
    (
        63,
        2,
        "(* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> 1))) (cast<int32> ?x1))",
    ),
    (
        64,
        2,
        "(* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> (neg (cast<int32> 1))))) (cast<int32> ?#1))",
    ),
    (
        67,
        2,
        "(* (cast<int32> (+ (cast<int32> ?#1) (cast<int32> ?x0))) (cast<int32> 4))",
    ),
    (
        75,
        2,
        "(<= (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        82,
        2,
        "(get ?#1 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 256))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        84,
        2,
        "(get ?#1 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 258))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        85,
        2,
        "(get ?#1 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 384))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        87,
        2,
        "(get ?#1 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 386))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        90,
        2,
        "(get ?#1 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?#2) (cast<int32> 128))) (cast<int32> ?x0))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        32,
        3,
        "(+ (cast<int32> (call[min] (cast<int32> (call[max] (cast<int32> ?x0) (cast<int32> 1))) (cast<int32> (+ (cast<int32> ?x0) (cast<int32> 6))))) (cast<int32> (neg (cast<int32> 1))))",
    ),
    (
        37,
        3,
        "(+ (cast<int32> ?x0) (cast<int32> 5))",
    ),
    (
        57,
        3,
        "(- (cast<int32> (call[min] (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        65,
        3,
        "(* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> (+ (cast<int32> ?#2) (cast<int32> 128))))",
    ),
    (
        70,
        3,
        "(* (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        73,
        3,
        "(/ (cast<int32> ?x0) (cast<int32> 128))",
    ),
    (
        77,
        3,
        "(get ?x0 (cast<int32x128> (call[ramp] (cast<int32> ?x1) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        79,
        3,
        "(get ?#1 (cast<int32x128> (+ (cast<int32x128> (call[max] (cast<int32x128> (call[min] (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (+ (cast<int32> (* (cast<int32> ?#2) (cast<int32> 128))) (cast<int32> ?#3))) (cast<int32> (neg (cast<int32> 1))))) (cast<int32> 1) (cast<int32> 128))) (cast<int32x128> (call[x128] (cast<int32> (+ (cast<int32> ?#4) (cast<int32> (neg (cast<int32> 1))))))))) (cast<int32x128> (call[x128] (cast<int32> 0))))) (cast<int32x128> (call[x128] (cast<int32> ?x0))))))",
    ),
    (
        34,
        4,
        "(+ (cast<int32> ?x0) (cast<int32> 1))",
    ),
    (
        48,
        4,
        "(+ (cast<uint16x128> (call[absd] (cast<uint16x128> (+ (cast<uint16x128> (call[widening_shift_left] (cast<uint8x128> (get ?#16 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> ?x1))) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> 1))))) (cast<uint16x128> (call[widening_add] (cast<uint8x128> ?x2) (cast<uint8x128> ?x3))))) (cast<uint16x128> (+ (cast<uint16x128> (call[widening_shift_left] (cast<uint8x128> (get ?#17 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x4) (cast<int32> 128))) (cast<int32> ?x5))) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> 1))))) (cast<uint16x128> (call[widening_add] (cast<uint8x128> ?x6) (cast<uint8x128> ?x7))))))) (cast<uint16x128> (call[absd] (cast<uint16x128> (+ (cast<uint16x128> (call[widening_shift_left] (cast<uint8x128> (get ?#18 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x8) (cast<int32> 128))) (cast<int32> ?x9))) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> 1))))) (cast<uint16x128> (call[widening_add] (cast<uint8x128> ?x2) (cast<uint8x128> ?x6))))) (cast<uint16x128> (+ (cast<uint16x128> (call[widening_shift_left] (cast<uint8x128> (get ?#19 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x8) (cast<int32> 128))) (cast<int32> ?x10))) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> 1))))) (cast<uint16x128> (call[widening_add] (cast<uint8x128> ?x3) (cast<uint8x128> ?x7))))))))",
    ),
    (
        68,
        4,
        "(* (cast<int32> ?x0) (cast<int32> 4))",
    ),
    (
        38,
        5,
        "(+ (cast<int32> ?x0) (cast<int32> 6))",
    ),
    (
        69,
        5,
        "(* (cast<int32> ?x0) (cast<int32> 128))",
    ),
    (
        27,
        7,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> ?x1))",
    ),
    (
        60,
        8,
        "(- (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        28,
        10,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        16,
        11,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 4))) (cast<int32> ?x1))",
    ),
    (
        45,
        20,
        "(+ (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
]
